### PR TITLE
feat(cache): RENOVATE_X_REPO_CACHE_FORCE_LOCAL

### PR DESCRIPTION
**Pull request by PhilipAbed**
## Changes

added a flag to persist the data in local cache after uploading to s3

## Context

see discussion
https://github.com/renovatebot/renovate/discussions/28217

## Documentation (please check one with an [x])

- [x] I have updated the documentation, or
- [ ] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/util/cache/repository/impl/s3.ts [hunk 1]
```diff
@@ -54,14 +56,22 @@ export class RepoCacheS3 extends RepoCacheBase {
 
   async write(data: RepoCacheRecord): Promise<void> {
     const cacheFileName = this.getCacheFileName();
+    const Body = JSON.stringify(data);
```
Comment (rarkins) [inline lib/util/cache/repository/impl/s3.ts:59]:
```suggestion
    const stringifiedCache = JSON.stringify(data);
```

File: lib/util/cache/repository/impl/s3.ts [hunk 2]
```diff
@@ -54,14 +56,22 @@ export class RepoCacheS3 extends RepoCacheBase {
...
     const s3Params: PutObjectCommandInput = {
       Bucket: this.bucket,
       Key: cacheFileName,
-      Body: JSON.stringify(data),
+      Body,
```
Comment (rarkins) [inline lib/util/cache/repository/impl/s3.ts:63]:
```suggestion
      Body: stringifiedCache,
```

File: lib/util/cache/repository/impl/s3.ts [hunk 3]
```diff
@@ -54,14 +56,22 @@ export class RepoCacheS3 extends RepoCacheBase {
...
       ContentType: 'application/json',
     };
     try {
       await this.s3Client.send(new PutObjectCommand(s3Params));
+      if (process.env.RENOVATE_X_REPO_CACHE_S3_LOCAL === 'true') {
+        const cacheLocalFileName = getLocalCacheFileName(
+          this.platform,
+          this.repository,
+        );
+        await outputCacheFile(cacheLocalFileName, Body);
```
Comment (rarkins) [inline lib/util/cache/repository/impl/s3.ts:73]:
```suggestion
        await outputCacheFile(cacheLocalFileName, stringifiedCache);
```

**Review by rarkins — CHANGES_REQUESTED** [review]

File: lib/util/cache/repository/impl/s3.ts [hunk 4]
```diff
@@ -54,14 +56,22 @@ export class RepoCacheS3 extends RepoCacheBase {
...
+      Body: stringifiedCache,
       ContentType: 'application/json',
     };
     try {
       await this.s3Client.send(new PutObjectCommand(s3Params));
+      if (process.env.RENOVATE_X_REPO_CACHE_S3_LOCAL === 'true') {
+        const cacheLocalFileName = getLocalCacheFileName(
+          this.platform,
+          this.repository,
+        );
```
Comment (rarkins) [inline lib/util/cache/repository/impl/s3.ts:72]: Is this the same file name as when repoCache is local not s3?

Comment (PhilipAbed) [reply lib/util/cache/repository/impl/s3.ts:72]: yes, this is the annoying  part, in s3 the file name is repo/cache.json
in local its repo.json

should i rename it with a replace?
or do i leave it with the same behaviour as local exactly?

Comment (rarkins) [reply lib/util/cache/repository/impl/s3.ts:72]: It should keep the same file name as local behavior

**Review by rarkins — CHANGES_REQUESTED** [review]
Change to RENOVATE_X_REPO_CACHE_FORCE_LOCAL

File: lib/util/cache/repository/impl/s3.ts (hunk 4 above)
Comment (rarkins) [reply lib/util/cache/repository/impl/s3.ts:72]: Is this resolved? If so how?

Comment (PhilipAbed) [reply lib/util/cache/repository/impl/s3.ts:72]: i extracted the getLocalCacheFileName function
and im using it in both `s3.ts` and `local.ts` to maintain same behaviour

no changes required.

**Review by rarkins — DISMISSED** [review]

**Review by viceice — DISMISSED** [review]

File: lib/util/cache/repository/impl/s3.ts [hunk 5]
```diff
@@ -54,14 +56,22 @@ export class RepoCacheS3 extends RepoCacheBase {
...
     const s3Params: PutObjectCommandInput = {
       Bucket: this.bucket,
       Key: cacheFileName,
-      Body: JSON.stringify(data),
+      Body: stringifiedCache,
       ContentType: 'application/json',
     };
     try {
       await this.s3Client.send(new PutObjectCommand(s3Params));
+      if (process.env.RENOVATE_X_REPO_CACHE_FORCE_LOCAL === 'true') {
```
Comment (secustor) [inline lib/util/cache/repository/impl/s3.ts:68]:
```suggestion
      if (is.nonEmptyString(process.env.RENOVATE_X_REPO_CACHE_FORCE_LOCAL)) {
```
This is currently not doing what is described in the docs. 

It expects 'true' rather checking for any setting. 
We should change the docs or how this is detected.

**Review by secustor — CHANGES_REQUESTED** [review]

**Review by secustor — APPROVED** [review]

**Review by rarkins — APPROVED** [review]

**Review by rarkins — APPROVED** [review]

**Review by viceice — APPROVED** [review]

**Pull request merged**